### ISSUE TITLE
Offer to write documentation

### ISSUE THREAD

**Issue by stnbu** [tomusdrw/rust-web3#505]
Hello!

I'm having a hard time finding documentation and/or examples for `rust-web3`. I'm new to rust, so I have that as a disadvantage (but I do bring fresh eyes.)

I see the `examples` directory, but as a rust newbie, I run into difficulties running these. I'm sure it's something simple, which I'd love to document or improve somehow to lower the barrier to folks like me.

I would be most grateful if you could respond with some documentation links in this issue. I've looked in all the usual places, but haven't had much luck. Many are out-of-date unfortunately (and depend on a github url instead of a version number, so one has to fiddle with versions to find the point where things break.) Is `rust-web3` used in other active projects? I can't find much of that either.

My end-goal is to write some code to compile to wasm and use in a web-hosted dapp. I'm doing this primarily as a means of learning rust+ethereum. I've read some issues here about compiling to wasm, and if I understand them correctly, I'm ok with the constraints (like maybe having to restrict use to metamask).

Can we trade? I get some help on getting bootstrapped with `rust-web3`, and in return I write _some_ docs...? ;-) If not that's ok too. Thanks Team.

**Comment by bytesoul** [comment]
Hi I am pretty new to this as well and have been working on something similar to your end goal. 

So far I have the web3 module compiling to wasm and wasm --target web and I am able to connect to a local geth instance.
Some notes 
I would start with the wasm-bindgen example https://github.com/rustwasm/wasm-bindgen/tree/master/examples/without-a-bundler

Add the web3 module in the cargo toml like this 
web3 = { version = "0.16.0", default-features = false, features = ["http","eip-1193","signing"] }

make sure you are using the 0.16.0 version and enable the wasm features (this can either be done with the eip-1193 flag or the wasm flag). Also to compile the signing feature you will have to have clang installed.

for an example I altered the README one like so 

pub async fn webexample() -> web3::Result<()>{

    let http = web3::transports::Http::new("http://127.0.0.1:8545")?; //local node
    let web3 = web3::Web3::new(http);
    let mut accounts = web3.eth().accounts().await?;
    alert(&format!("Hello, {:?}!", accounts));

    accounts.push("680640838Bd07A447B168d6D923b90CF6c43CDcA".parse().unwrap()); // random account 
    alert(&format!("Hello, {:?}!", accounts));

    for account in accounts {
        let balance = web3.eth().balance(account, None).await?;
        alert(&format!("Balance of {:?}: {}", account, balance));
    }

    Ok(())

}

I tried to import this function using [#wasm_bindgen] but the return type of web3::Result is not compatible some one more experienced may be able to help. But you can call the function in your rust code in the main and the alerts will pop up. Connecting to metamask or other endpoints I have not tested.

Hope this helps.

**Comment by stnbu** [comment]
I've since got sidetracked to `ethers-rs` but I'm still figuring things out there obviously. I'm sure I'll circle back here since there is more talk of wasm support in this project. I've also figured out my issues and probably could make `rust-web3` work for me now. I'm not sure what I was doing wrong. It had to do with some or all of `tokio` missing but I think I've since found the right `Cargo.toml` settings. For [ethers](https://github.com/stnbu/sandbox/blob/d76306d9f517ff63806bff26016628a51d1a84fc/rust/contraction/Cargo.toml) at least. (no wasm but a minimal, complete example of using `ethers-rs` to read a value from a contract.)

**Comment by tomusdrw** [comment]
> I tried to import this function using [#wasm_bindgen] but the return type of web3::Result is not compatible some one more experienced may be able to help.

@bytesoul The `web3::Result` return type of `main` is there for convenience so that we can use `?` in the code. You could simply just unwrap or handle the last error and return something wasm-compatible.


@stnbu More than happy to merge more sophisticated examples and a base crate setup. So far the examples in `example` directory seemed to be sufficient, but yeah, the `Cargo.toml` setup was assumed to be something that people can deal with on their own. As I've mentioned happy to merge an "example crate" which will contain WASM instructions and overall project setup.

Regarding the outdated documentation links, was it something you've found in this repository? I'm also more than happy to update them given the pointers.

**Comment by stnbu** [comment]
> Regarding the outdated documentation links, was it something you've found in this repository? I'm also more than happy to update them given the pointers.

No, this was in an online "tutorial" type thing. Medium.com or similar. Thanks for the response. I'm still doing a lot of "circling" but I definitely want to contribute. Having learned a bit more, `ethers-rs` still appears to be less far along than `rust-web3` in regards to wasm support.

**Comment by isim95** [comment]
hey @stnbu, I am in a similar situation, new to learning Rust + Solidity by creating a dapp based in rust, but am having trouble getting some of the exmaples to run (in my case I'm trying to figure out how to listen to a particular contract's transactions for each new block)

**Comment by stnbu** [comment]
Hey, @isim95 I invited you to [abandoned-addresses](https://github.com/stnbu/abandoned-addresses) which is a sort of an otherwise pointless learn-by-doing project I periodically use for stuff like this. My goal is: replace the ethers.js stuff with Rust WASM. I realize that that _might_ be impractical or a bad idea for other reasons, but it's just a pallet for me to practice at Rust. And I can imagine generally "web3 capable wasm" will be a thing some day. Might even have a practical use.

Feel free to ignore. But you're also welcome to throw whatever you like at it (I'll be creating the crate for the above today, so just starting).

[Sorry @tomusdrw for using your project's issues as my personal discussion forum. I'm pretty sure I have good intentions. Sorry github for eating all your bits.]
